pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
 // lists should be merged or replaced.
 //
 // For more information, see the PATCH section of docs/api-conventions.md.
+
+// CreateStrategicMergePatch creates a patch that can be passed to StrategicMergePatch.
```

Comment:
Rather than bail if we detect any divergence between the original configuration to the live state, I think we want to bail only if we detect conflicts between such a divergence and the patch that the user now wants to apply. That's really the point of generating the strategic merge patch, right? We can apply the changes that the user is requesting on top of changes made by automated processes, if and only if there are no conflicts between the two change sets.

Ignoring for the moment the possibility of conflicts between different but related properties, we should bail if cases where the same properties are changed in conflicting ways. In other words, we should bail if there are any properties in both change sets that have different values. We're fine if one change set changed a property that the other change set didn't touch, and we're fine if both change sets changed the same property to the same value.

The OpenShift json merge implementation detects these simple conflicts by generating the patch required to go from original to live, and then comparing it with the patch being applied. However, it doesn't attempt to ignore additions, since they don't generate conflicts anyway. In other words, we didn't really the flag to ignore additions. If we don't have a compelling use case for it, then I think it would be cleaner to back it out and always generate a complete patch that includes additions, as well as deletions and modifications.

The OpenShift json merge implementation also provides a mechanism for detecting more complex conflicts involving incompatible changes to related properties: it allows the caller to register a set of preconditions and then plays them against any given patch before applying it. If the caller knows about constraints that span two or more properties, then they can write preconditions to look for violations of those constraints, and register them with the library. I'm inclined to include this feature, since it's small and simple, and provides significant power.